core/src/main/java/org/infinispan/interceptors/locking/PessimisticLockingInterceptor.java

```diff
@@ -232,9 +232,7 @@ private boolean readNeedsLock(InvocationContext ctx, LocalFlagAffectedCommand co
    @Override
    public CompletableFuture<Void> visitApplyDeltaCommand(InvocationContext ctx, ApplyDeltaCommand command) throws Throwable {
       ctx.onReturn((rCtx, rCommand, rv, throwable) -> {
-         if (throwable != null) {
-            lockManager.unlockAll(rCtx);
-         }
+         rethrowAndReleaseLocksIfNeeded(rCtx, throwable);
```

Comment:
@pruivo Please make this change for read commands as well - `StateTransferInterceptor` also catches `OutdatedTopologyException` and retries for them, even though not not all of them can throw `OTE`.